### DISCUSSION TITLE
Is it able to build a parse header only gateway using hyper?

### DISCUSSION THREAD

**Discussion by Dennis-Zhang-SH** [hyperium/hyper#3186] [General]
Hi, I am a newbie to rust and hyper, is it able that I only parse the header part of the request, and validate the request then forward it?

**Comment by seanmonstar** [comment]
Based on your short description, it sounds like yes. The gateway example does that.

**Reply by Dennis-Zhang-SH** [reply]
> Based on your short description, it sounds like yes. The gateway example does that.

Hi, I just checked gateway example, I am assuming that `req` in `service_fn` is a completely parsed request? Actually what I want is parsing the header at first, get the authentication token in the header, check if it's valid, if so keep parsing and forward the request, If not stop parsing and return with 401 error.

**Reply by seanmonstar** [reply]
hyper parses the request headers, the body is not read yet. The body type is a handle to let you stream the body when you want.

**Reply by Dennis-Zhang-SH** [reply]
> hyper parses the request headers, the body is not read yet. The body type is a handle to let you stream the body when you want.

Hi, I tried the gateway example, but it turns out under concurrent pressure, the result is not looking good.

![image](https://user-images.githubusercontent.com/51769674/228102993-3185ac6c-bae6-49d2-8cef-9f5bcdafd47d.png)

At port 3000, it's just simple axum example code:

![image](https://user-images.githubusercontent.com/51769674/228103104-cab10a29-6b9e-4823-a4da-188964e5d9e5.png)

Here is part of log:

```log
Listening on http://127.0.0.1:3001
Proxying on http://127.0.0.1:3000
Failed to servce connection: hyper::Error(IncompleteMessage)
Failed to servce connection: hyper::Error(IncompleteMessage)
Failed to servce connection: hyper::Error(IncompleteMessage)
Failed to servce connection: hyper::Error(Shutdown, Os { code: 107, kind: NotConnected, message: "Transport endpoint is not connected" })
Failed to servce connection: hyper::Error(Shutdown, Os { code: 107, kind: NotConnected, message: "Transport endpoint is not connected" })
Failed to servce connection: hyper::Error(Shutdown, Os { code: 107, kind: NotConnected, message: "Transport endpoint is not connected" })
Failed to servce connection: hyper::Error(Shutdown, Os { code: 107, kind: NotConnected, message: "Transport endpoint is not connected" })
Failed to servce connection: hyper::Error(IncompleteMessage)
Failed to servce connection: hyper::Error(Shutdown, Os { code: 107, kind: NotConnected, message: "Transport endpoint is not connected" })
Failed to servce connection: hyper::Error(Shutdown, Os { code: 107, kind: NotConnected, message: "Transport endpoint is not connected" })
Failed to servce connection: hyper::Error(Shutdown, Os { code: 107, kind: NotConnected, message: "Transport endpoint is not connected" })
Failed to servce connection: hyper::Error(Shutdown, Os { code: 107, kind: NotConnected, message: "Transport endpoint is not connected" })
Failed to servce connection: hyper::Error(Shutdown, Os { code: 107, kind: NotConnected, message: "Transport endpoint is not connected" })
```

While using other simple proxy like `caddy`, the result looks much better:

![image](https://user-images.githubusercontent.com/51769674/228103865-7424a66e-3cfa-4e08-a605-d5979fe9e7b3.png)

Would you mind explaining why this happened?
